### ISSUE TITLE
'Broken pipe' messages logged during read load

### ISSUE THREAD

**Issue by pdziepak** [scylladb/scylla#2738]
*Installation details*
Scylla version (or git commit hash): 2.0-rc2
Cluster size: 3x i2.8xlarge
Loaders: 10x c4.4xlarge

Scylla was populated with the following load run for 1h and then for 30min.
```
scylla-bench -duration 1h -partition-count 1000000 -clustering-row-count 1000 -clustering-row-size 128 -connection-count 32 -measure-latency=false -replication-factor 3 -concurrency=256 -workload=uniform -mode=write
```

Then, all nodes were restarted and the following read load was started:

```
scylla-bench -duration 2h -partition-count 1000000 -clustering-row-count 1000 -clustering-row-size 128 -connection-count 32 -measure-latency=false -replication-factor 3 -concurrency=256 -workload=uniform -mode=read
```

As the cache warmed up some of nodes (in one run all, but there were runs that only one) start logging the following message:
```
Aug 16 17:42:03 ip-172-30-0-39 scylla[42419]:  [shard 0] seastar - Exceptional future ignored: std::system_error (error system:32, Broken pipe), backtrace: 0x4f0123, 0x4f0181, 0x58a3af, 0x90f989, 0x4ecd4f, 0x54087a, 0x5c4fb4, 0x47524f, 0x7f5a8810eb14, 0x4e8dde
```

The backtrace translates to:
```
seastar::current_backtrace() at /usr/src/debug/scylla-2.0.rc2/seastar/core/reactor.cc:3935
seastar::report_failed_future(std::__exception_ptr::exception_ptr) at ??:?
seastar::future<>::~future() at /usr/src/debug/scylla-2.0.rc2/seastar/core/future.hh:799
 (inlined by) std::_Head_base<1ul, seastar::future<>, false>::~_Head_base() at /opt/scylladb/include/c++/5.3.1/tuple:102
_ZN7seastar12continuationIZNS_6futureIISt5tupleIINS1_IIEEES3_EEEE4thenIZNS_5httpd10connection7processEvEUlS4_E_S3_EET0_OT_EUlSC_E_IS4_EE3runEv at ??:?
seastar::reactor::run_tasks(seastar::circular_buffer<std::unique_ptr<seastar::task, std::default_delete<seastar::task> >, std::allocator<std::unique_ptr<seastar::task, std::default_delete<seastar::task> > > >&) at ??:?
seastar::reactor::run() at ??:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at /usr/src/debug/scylla-2.0.rc2/seastar/core/app-template.cc:142
main at ??:?
?? ??:0
_start at ??:?
```

**Comment by pdziepak** [comment]
Possibly related to the stalls reported in #2698.

**Comment by amnonh** [comment]
I've tried to reproduce the problem with rc4 multiple times and couldn't.

**Comment by slivne** [comment]
closing his issue - if it does reproduce we can reopen

**Comment by tgrabiec** [comment]
@amnonh Were you able to reproduce the problem on rc2, on which the issue was reported?

**Comment by amoskong** [comment]
@slivne I just reproduced this issue in longevity test with `2.0.rc5-0.20170917.0b49cfc`.

Monitor: http://35.196.66.85:3000/dashboard/db/scylla-per-server-metrics-2-0?from=1506200413796&to=1506206222165
Find the instances in GCE by search `longevity-50gb-4d-20rc5-amos`

Test context around the error:
- decommission `longevity-50gb-4d-20rc5-amos-db-node-3ad7b057-0-2` (10.240.0.61)
- 2017-09-23 21:40:15,092 `nodetool decommission` costed about 20 mins (finished, good).
- 2017-09-23 21:41:33,531 destroy instance longevity-50gb-4d-20rc5-amos-db-node-3ad7b057-0-2
- 2017-09-23 21:41:50,685   `longevity-50gb-4d-20rc5-amos-db-node-3ad7b057-2-13` [10.240.0.89] |  [shard 0] seastar - Exceptional future ignored: std::system_error (error system:32, Broken pipe), backtrace: 0x4f0073, 0x4f00d1, 0x58a2ff, 0x90fc79, 0x4ecc9f, 0x5407ca, 0x5c4f04, 0x47514f, 0x7fbceb10bc04, 0x4e8d2e
- 2017-09-23 21:41:50,705   `longevity-50gb-4d-20rc5-amos-db-node-3ad7b057-1-9` 
 [10.240.0.83]  [shard 0] seastar - Exceptional future ignored: std::system_error (error system:32, Broken pipe), backtrace: 0x4f0073, 0x4f00d1, 0x58a2ff, 0x90fc79, 0x4ecc9f, 0x5407ca, 0x5c4f04, 0x47514f, 0x7f7f80782c04, 0x4e8d2e
- 2017-09-23 21:53:33,902 start new node
- new node is waiting `4+ hours` for schema information to complete, so it's not up
- 2017-09-24 00:48:16,402 remote           L0197 DEBUG| [10.240.0.5] [stdout] Sep 24 00:48:16 info   |  [shard 0] storage_service - JOINING: waiting for schema information to complete

Environment:
- Version: scylla-server-2.0.rc5-0.20170917.0b49cfc.el7.centos.x86_64
- 3 datacenters (us-east1-b, us-west1-b, us-central1-b)
- each datacenter has 5 db nodes (n1-highmem-8)
- 3 loaders, one loader for one dc
- workloads：write + read + mv
```
cassandra-stress read cl=QUORUM duration=5760m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=500 -pop seq=1..100000000 -log interval=5
"cassandra-stress write cl=QUORUM duration=5760m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=1000 -pop seq=1..100000000 -log interval=5",
"cassandra-stress counter_write cl=QUORUM duration=5760m -schema 'replication(factor=3) compaction(strategy=DateTieredCompactionStrategy)' -port jmx=6868 -mode cql3 native -rate threads=1000 -pop seq=1..1000000",
"cassandra-stress user profile=/tmp/cs_mv_profile.yaml ops'(insert=3,read1=1,read2=1,read3=1)' cl=QUORUM duration=5760m -port jmx=6868 -mode cql3 native -rate threads=100
```

```
[scylla-test@longevity-50gb-4d-20rc5-amos-db-node-3ad7b057-1-9 ~]$ addr2line -Cfpi -e /usr/lib/debug/usr/bin/scylla.debug 0x4f0073, 0x4f00d1, 0x58a2ff, 0x90fc79, 0x4ecc9f, 0x5407ca, 0x5c4f04, 0x47514f, 0x7fbceb10bc04, 0x4e8d2e
seastar::current_backtrace() at /usr/src/debug/scylla-2.0.rc5/seastar/core/reactor.cc:3935
seastar::report_failed_future(std::__exception_ptr::exception_ptr) at /usr/src/debug/scylla-2.0.rc5/seastar/core/reactor.cc:3960
seastar::future<>::~future() at /usr/src/debug/scylla-2.0.rc5/seastar/core/future.hh:799
 (inlined by) std::_Head_base<1ul, seastar::future<>, false>::~_Head_base() at /opt/scylladb/include/c++/5.3.1/tuple:102
std::_Tuple_impl<1ul, seastar::future<> >::~_Tuple_impl() at /opt/scylladb/include/c++/5.3.1/tuple:339
 (inlined by) std::_Tuple_impl<0ul, seastar::future<>, seastar::future<> >::~_Tuple_impl() at /opt/scylladb/include/c++/5.3.1/tuple:180
 (inlined by) std::tuple<seastar::future<>, seastar::future<> >::~tuple() at /opt/scylladb/include/c++/5.3.1/tuple:596
 (inlined by) seastar::apply_helper<seastar::httpd::connection::process()::{lambda(std::tuple<seastar::future<>, seastar::future<> >)#1}, std::tuple<std::tuple<seastar::future<>, seastar::future<> > >&&, std::integer_sequence<unsigned long, 0ul> >::apply({lambda(std::tuple<seastar::future<>, seastar::future<> >)#1}&&, std::tuple<std::tuple<seastar::future<>, seastar::future<> > >) at /usr/src/debug/scylla-2.0.rc5/seastar/core/apply.hh:36
 (inlined by) auto seastar::apply<seastar::httpd::connection::process()::{lambda(std::tuple<seastar::future<>, seastar::future<> >)#1}, std::tuple<seastar::future<>, seastar::future<> > >(seastar::httpd::connection::process()::{lambda(std::tuple<seastar::future<>, seastar::future<> >)#1}&&, std::tuple<std::tuple<seastar::future<>, seastar::future<> > >&&) at /usr/src/debug/scylla-2.0.rc5/seastar/core/apply.hh:44
 (inlined by) seastar::future<> seastar::futurize<seastar::future<> >::apply<seastar::httpd::connection::process()::{lambda(std::tuple<seastar::future<>, seastar::future<> >)#1}, std::tuple<seastar::future<>, seastar::future<> > >(seastar::httpd::connection::process()::{lambda(std::tuple<seastar::future<>, seastar::future<> >)#1}&&, std::tuple<std::tuple<seastar::future<>, seastar::future<> > >&&) at /usr/src/debug/scylla-2.0.rc5/seastar/core/future.hh:1302
 (inlined by) _ZZN7seastar6futureIJSt5tupleIJNS0_IJEEES2_EEEE4thenIZNS_5httpd10connection7processEvEUlS3_E_S2_EET0_OT_ENUlSB_E_clINS_12future_stateIJS3_EEEEEDaSB_ at /usr/src/debug/scylla-2.0.rc5/seastar/core/future.hh:900
 (inlined by) _ZN7seastar12continuationIZNS_6futureIJSt5tupleIJNS1_IJEEES3_EEEE4thenIZNS_5httpd10connection7processEvEUlS4_E_S3_EET0_OT_EUlSC_E_JS4_EE3runEv at /usr/src/debug/scylla-2.0.rc5/seastar/core/future.hh:395
seastar::reactor::run_tasks(seastar::circular_buffer<std::unique_ptr<seastar::task, std::default_delete<seastar::task> >, std::allocator<std::unique_ptr<seastar::task, std::default_delete<seastar::task> > > >&) at /usr/src/debug/scylla-2.0.rc5/seastar/core/reactor.cc:2316
seastar::reactor::run() at /usr/src/debug/scylla-2.0.rc5/seastar/core/reactor.cc:2774
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at /usr/src/debug/scylla-2.0.rc5/seastar/core/app-template.cc:142
main at /usr/src/debug/scylla-2.0.rc5/main.cc:682
?? ??:0
_start at ??:?
```
20mins snapshot around the error:
![longevity-mdc-20mins-range](https://user-images.githubusercontent.com/309708/30778561-5fb9f2a4-a10b-11e7-943f-9ac2d28077d1.png)

2hours snapshot around the error:
[longevity-mdc-2hours-range](https://user-images.githubusercontent.com/309708/30778563-65af500a-a10b-11e7-9f76-42d3778b17cb.png)

Full snapshot (9+ hours):
[longevity-mdc-full-range](https://user-images.githubusercontent.com/309708/30778564-69a015a0-a10b-11e7-9ee3-fe9021eb6c93.png)

**Comment by slivne** [comment]
@amnonh - please look at this

**Comment by amnonh** [comment]
I've created a small python script that read from the socket, timeout and disconnect.
I verify that the error metrics in the http report it, but there is no broken pipe, so the main code path is working as expected.

looking for other things that can go wrong

**Comment by gleb-cloudius** [comment]
On Sun, Sep 24, 2017 at 11:24:55AM +0000, Amnon Heiman wrote:
> I've created a small python script that read from the socket, timeout and disconnect.
> I verify that the error metrics in the http report it, but there is no broken pipe, so the main code path is working as expected.
> 
> looking for other things that can go wrong
> 
b9bb16413c02c6e92b12f746669602a24d8b0bb3 is not in scylla-2.0

--
			Gleb.

**Comment by amnonh** [comment]
it probably should, but b9bb16413c02c6e92b12f746669602a24d8b0bb3 is during shutdown, I don't think it impact the normal path

**Comment by gleb-cloudius** [comment]
On Sun, Sep 24, 2017 at 01:23:08PM +0000, Amnon Heiman wrote:
> it probably should, but b9bb16413c02c6e92b12f746669602a24d8b0bb3 is during shutdown, I don't think it impact the normal path
> 
It impacts socket closer. The trace is exactly at the place where
ignore_ready_future() were added. The commit log is misleading.

--
			Gleb.

**Comment by amnonh** [comment]
On Sun, Sep 24, 2017 at 4:26 PM, Gleb Natapov <notifications@github.com>
wrote:

> On Sun, Sep 24, 2017 at 01:23:08PM +0000, Amnon Heiman wrote:
> > it probably should, but b9bb16413c02c6e92b12f746669602a24d8b0bb3 is
> during shutdown, I don't think it impact the normal path
> >
> It impacts socket closer. The trace is exactly at the place where
> ignore_ready_future() were added. The commit log is misleading.
>
I'm looking at the code now, I think you are right and it can has an effect
(which probably why I'm unable to reproduce it locally no matter what I do
to the system),
I'm thinking that maybe I should change the code there from then to
then_wrapped.

@avi can you take a look at the commit and see if we can merge it?

specifically, the part under connection::process

**Comment by gleb-cloudius** [comment]
On Sun, Sep 24, 2017 at 06:36:51AM -0700, Amnon Heiman wrote:
> On Sun, Sep 24, 2017 at 4:26 PM, Gleb Natapov <notifications@github.com>
> wrote:
> 
> > On Sun, Sep 24, 2017 at 01:23:08PM +0000, Amnon Heiman wrote:
> > > it probably should, but b9bb16413c02c6e92b12f746669602a24d8b0bb3 is
> > during shutdown, I don't think it impact the normal path
> > >
> > It impacts socket closer. The trace is exactly at the place where
> > ignore_ready_future() were added. The commit log is misleading.
> >
> I'm looking at the code now, I think you are right and it can has an effect
> (which probably why I'm unable to reproduce it locally no matter what I do
> to the system),
> I'm thinking that maybe I should change the code there from then to
> then_wrapped.
> 
You cannot since when_all() always return valid future that contain a
tuple that may contain failed future. If you'll replace when_all() with
when_all_succeed() you can use handle_exception([](auto ex){}) there.

--
			Gleb.

**Comment by amnonh** [comment]
I have tested and verified that that commit b9bb16413c02c6e92b12f746669602a24d8b0bb3 does solve the problem.

We should probably backport it

**Comment by tgrabiec** [comment]
Fixes by dcac7125feb5a958b11b259481ac4d66aefc4103
